File: erlang/seminars/shoutcast/src/http.erl
-module(http).

-export([test/0, parse_get/1, request_line/1]).

test() ->
    parse_get("GET /foo HTTP/1.1\r\nfoo: 17\r\nbar: 42\r\n\r\nThis is the body").

parse_get(Request) ->
    {ok, Method, Resource, Version, R1} = request_line(Request),
    {ok, Headers, Body} = headers(R1),
    {ok, Method, Resource, Version, Headers, Body}.
    

headers([13,10|Rest]) ->
    {ok, [], Rest};
headers(Headers) ->
    {ok, Header, R1} = header(Headers),
    {ok, Rest, R2} = headers(R1),
    {ok, [Header|Rest], R2}.


header([13,10|Rest]) ->
    {ok, [], Rest};
header([C|Rest]) ->
    {ok, Header, R1} = header(Rest),
    {ok, [C|Header], R1}.


request_line([$G,$E,$T, 32 | R0]) ->
    {Resource, R1} = resource(R0),
    {Version, R2} = version(R1),
    [13,10|R3] = R2,
    {ok, get, Resource, Version, R3}.

resource([32|Rest]) ->
    {[], Rest};
resource([C|Rest]) ->
    {Resource, Cont} = resource(Rest),
    {[C|Resource], Cont}.


%% NOT TRUE but ok, general form is HTTP/12.34 etc

version([$H,$T,$T,$P, $/, $1, $., $2 | Rest]) ->
    {version12, Rest};
version([$H,$T,$T,$P, $/, $1, $., $1 | Rest]) ->
    {version11, Rest};
version([$H,$T,$T,$P, $/, $1, $., $0 | Rest]) ->
    {version10, Rest}.

File: examples/parser/http.ex
defmodule HTTP do

  def parse(r0) do
    {request, r1} = request(r0)
    {headers, r2} = headers(r1)
    {body, _} = body(r2)
    {request, headers, body}
  end

  
  
  def request([?G, ?E, ?T, 32 | r0]) do
    {uri, r1} = uri(r0)
    {ver, r2} = version(r1)
    [13, 10 | r3] = r2
    {{:get, uri, ver}, r3}
  end

  def uri([32 | r0]), do: {[], r0}
  def uri([c | r0]) do
    {rest, r1} = uri(r0)
    {[c | rest], r1}
  end

  def version([?H, ?T, ?T, ?P, ?/, ?1, ?., ?1 | r0]) do
    {:v11, r0}
  end
  def version([?H, ?T, ?T, ?P, ?/, ?1, ?., ?0 | r0]) do
    {:v10, r0}
  end

  def headers([13, 10 | r0]), do: {[], r0}
  def headers(r0) do
    {header, r1} = header(r0)
    {rest, r2} = headers(r1)
    {[header | rest], r2}
  end

  def header([13, 10 | r0]), do: {[], r0}
  def header([c | r0]) do
    {rest, r1} = header(r0)
    {[c | rest], r1}
  end

  def body(r), do: {r, []}

  

end  
